docs/source-fabric/advanced/model_parallel/tp.rst

```diff
@@ -0,0 +1,280 @@
+
+        import lightning as L
+        from lightning.pytorch.demos.boring_classes import RandomDataset
+        from lightning.fabric.strategies import ModelParallelStrategy
+
+
+        class FeedForward(nn.Module):
+            def __init__(self, dim, hidden_dim):
+                super().__init__()
+                self.w1 = nn.Linear(dim, hidden_dim, bias=False)
+                self.w2 = nn.Linear(hidden_dim, dim, bias=False)
+                self.w3 = nn.Linear(dim, hidden_dim, bias=False)
+
+            def forward(self, x):
+                return self.w2(F.silu(self.w1(x)) * self.w3(x))
+
+
+        def parallelize_feedforward(model, device_mesh):
+            # Lightning will set up a device mesh for you
+            tp_mesh = device_mesh["tensor_parallel"]
+            # Use PyTorch's distributed tensor APIs to parallelize the model
+            plan = {
+                "w1": ColwiseParallel(),
+                "w2": RowwiseParallel(),
+                "w3": ColwiseParallel(),
+            }
+            parallelize_module(model, tp_mesh, plan)
+            return model
+
+
+        strategy = ModelParallelStrategy(parallelize_fn=parallelize_feedforward)
+        fabric = L.Fabric(accelerator="cuda", devices=2, strategy=strategy)
+        fabric.launch()
+
+        # Initialize the model
+        model = FeedForward(8192, 8192)
+        model = fabric.setup(model)
+
+        # Define the optimizer
+        optimizer = torch.optim.AdamW(model.parameters(), lr=3e-3, foreach=True)
+        optimizer = fabric.setup_optimizers(optimizer)
+
+        # Define dataset/dataloader
+        dataset = RandomDataset(8192, 64)
+        dataloader = torch.utils.data.DataLoader(dataset, batch_size=8)
+        dataloader = fabric.setup_dataloaders(dataloader)
+
+        # Simplified training loop
+        for i, batch in enumerate(dataloader):
+            output = model(batch)
+            loss = output.sum()
+            fabric.backward(loss)
+            optimizer.step()
+            optimizer.zero_grad()
+            fabric.print(f"Iteration {i} complete")
+
+        fabric.print(f"Peak memory usage: {torch.cuda.max_memory_reserved() / 1e9:.02f} GB")
+
+|
+
+When measuring the peak memory consumption, we should see that doubling the number of GPUs reduces the memory consuption roughly by half:
+
+
+.. list-table::
+   :widths: 20 20 20 20 20
+   :header-rows: 1
+
+   * -
+     - 1 GPU (no TP)
+     - 2 GPUs
+     - 4 GPUs
+     - 8 GPUs
+   * - Memory per GPU
+     - 4.05 GB
+     - 2.15 GB
+     - 1.08 GB
+     - 0.61 GB
+
+Beyond this toy example, we recommend you study our `LLM Tensor Parallel Example (Llama 2) <https://github.com/Lightning-AI/pytorch-lightning/tree/master/examples/fabric/tensor_parallel>`_.
+
+
+----
+
+
+***************************
+Data-loading considerations
+***************************
+
+In a tensor-parallelized model, it is important that the model receives an identical input on each GPU.
```

Comment:
Intuitively, I would then think it can't be combined with data parallelism. But below you mention it can be combined, which is cool. Maybe you can say a few words here to clarify that it can still be used with data parallelism, and in short how it is handled (if it can be said in short).

Reply:
I personally think this is ok. At this point, the tutorial is focused on TP alone and I wanted to emphasize on this detail. In the future, I also want to write a section here about inference, where of course data-parallel is out of question anyway.
To explain how TP + DP is combined, I would rather rely on the separate page where we link (submitting a PR shortly) to fully explain it in detail.